streams/src/main/java/org/apache/kafka/streams/kstream/internals/InternalStreamsBuilder.java

```diff
@@ -252,6 +288,111 @@ public void buildAndOptimizeTopology() {
+                                                               parentSourceNode.keySerde(),
+                                                               parentSourceNode.valueSerde());
+
+            // re-use parent id to make sure the single repartition graph node is evaluated before downstream nodes
+            optimizedSingleRepartition.setId(parentSourceNode.id());
+
+            for (final OptimizableRepartitionNode repartitionNodeToBeReplaced : streamsGraphNodeSetEntry.getValue()) {
+
+                final StreamsGraphNode keyChangingNodeChild = findParentNodeMatching(repartitionNodeToBeReplaced, gn -> gn.parentNodes().contains(keyChangingNode));
+
+                LOG.debug("Found the node downstream of key-changer {}", keyChangingNodeChild);
+
+                // need to add children of key-changing node as children of optimized repartition
+                // in order to process records from re-partitioning
+                optimizedSingleRepartition.addChildNode(keyChangingNodeChild);
+
+                LOG.debug("Removing {} from {}  children {}", keyChangingNodeChild, keyChangingNode, keyChangingNode.children());
+                // now remove children from key-changing node
+                keyChangingNode.removeChild(keyChangingNodeChild);
+
+                // now need to get children of repartition node so we can remove repartition node
+                final Collection<StreamsGraphNode> repartitionNodeToBeReplacedChildren = repartitionNodeToBeReplaced.children();
+
+                final StreamsGraphNode parentOfRepartitionNodeToBeReplaced = findParentNodeMatching(repartitionNodeToBeReplaced, n -> repartitionNodeToBeReplaced.parentNodes().contains(n));
+
+                if (parentOfRepartitionNodeToBeReplaced == null) {
+                    throw new IllegalStateException(String.format("Can't find parent node for %s this should not happen and indicates a bug", keyChangingNode));
+                }
+
+                LOG.debug("Returned {} parentOfRepartitionNodeToBeReplaced as parent of {}", parentOfRepartitionNodeToBeReplaced, repartitionNodeToBeReplaced);
+
+                // remove repartition node
+                parentOfRepartitionNodeToBeReplaced.removeChild(repartitionNodeToBeReplaced);
+
+                repartitionNodeToBeReplaced.clearChildren();
+
+                // re-attach children of removed repartition
+                for (final StreamsGraphNode repartitionNodeToBeReplacedChild : repartitionNodeToBeReplacedChildren) {
+                    parentOfRepartitionNodeToBeReplaced.addChildNode(repartitionNodeToBeReplacedChild);
+                }
+
+                LOG.debug("Updated node {} children {}", optimizedSingleRepartition, optimizedSingleRepartition.children());
+            }
+
+            keyChangingNode.addChildNode(optimizedSingleRepartition);
+            keyChangingOperationsToOptimizableRepartitionNodes.remove(streamsGraphNodeSetEntry.getKey());
+        }
+    }
```

Comment:
I golfed this method a little to generate my other comments.

I was thinking you could reorganize it something like this:
```java
private void maybeOptimizeRepartitionOperations() {
    for (final Map.Entry<StreamsGraphNode, Set<OptimizableRepartitionNode>> entry : keyChangingOperationsToOptimizableRepartitionNodes.entrySet()) {
        final StreamsGraphNode keyChangingNode = entry.getKey();
        final Set<OptimizableRepartitionNode> optimizableRepartitionNodes = entry.getValue();

        if (optimizableRepartitionNodes.isEmpty()) {
            continue;
        }

        // Step 1: Create a new repartition node
        // Find the nearest upstream StreamSourceNode, which we know contains the right key/value serdes
        final StreamSourceNode parentSourceNode = (StreamSourceNode) findParentNodeMatching(keyChangingNode, n -> n instanceof StreamSourceNode);

        if (parentSourceNode == null) {
            throw new IllegalStateException(String.format("Can't find parent source node for %s ", keyChangingNode));
        }

        final OptimizableRepartitionNode optimizedSingleRepartition = createRepartitionNode(keyChangingNode.nodeName(),
                                                                                            parentSourceNode.keySerde(),
                                                                                            parentSourceNode.valueSerde());

        // Re-use parent priority to make sure the single repartition graph node is evaluated before downstream nodes
        optimizedSingleRepartition.setPriority(parentSourceNode.priority());

        // Step 2: Insert the new repartition node directly under the key-changing node
        for (final StreamsGraphNode keyChangingNodeChild : keyChangingNode.children()) {
            LOG.debug("Found the node downstream of key-changer {}", keyChangingNodeChild);

            // need to add children of key-changing node as children of optimized repartition
            // in order to process records from re-partitioning
            optimizedSingleRepartition.addChildNode(keyChangingNodeChild);

            LOG.debug("Removing {} from {}  children {}", keyChangingNodeChild, keyChangingNode, keyChangingNode.children());
            // now remove children from key-changing node
            keyChangingNode.removeChild(keyChangingNodeChild);
        }

        if (!keyChangingNode.children().isEmpty()) {
            throw new IllegalStateException();
        }

        keyChangingNode.addChildNode(optimizedSingleRepartition);

        // Step 3: Remove each of the old (now redundant) repartition nodes, hooking its children up to its parents
        for (final OptimizableRepartitionNode repartitionNodeToBeReplaced : optimizableRepartitionNodes) {
            // now need to get children of repartition node so we can remove repartition node
            final Collection<StreamsGraphNode> repartitionNodeToBeReplacedChildren = repartitionNodeToBeReplaced.children();

            for (final StreamsGraphNode parentNode : repartitionNodeToBeReplaced.parentNodes()) {
                parentNode.removeChild(repartitionNodeToBeReplaced);
                for (final StreamsGraphNode childNode : repartitionNodeToBeReplaced.children()) {
                    parentNode.addChildNode(childNode);
                }
            }

            repartitionNodeToBeReplaced.clearChildren();
        }

        keyChangingOperationsToOptimizableRepartitionNodes.remove(entry.getKey());
    }
}
```

WDYT?

Reply:
I've made some of the suggested changes, but I've left the inner loop as it's required and explained up above.

Reply:
For reasons mentioned above, I'd prefer to keep the traversal up the graph for child nodes of key changing operation so we can be more selective with the optimization operation. 

I did apply your suggestion for generalization for the removal of the repartition nodes which is much cleaner.